File: examples/CsharpExamples/Schema.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.Threading.Tasks;
using Microsoft.FSharp.Core;
using Pulsar.Client.Api;

namespace CsharpExamples
{
    public class Product
    {
        [Required]
        public string Name { get; set; }
        public double Rating { get; set; }
        public override string ToString()
        {
            return $"Name: {Name} Rating: {Rating}";
        }
    }
    
    internal class SchemaExample
    {
        internal static async Task RunSchema()
        {
            const string serviceUrl = "pulsar://my-pulsar-cluster:30002";
            const string subscriptionName = "my-subscription";
            var topicName = $"my-topic-{DateTime.Now.Ticks}";

            var client = await new PulsarClientBuilder()
                .ServiceUrl(serviceUrl)
                .BuildAsync();

            var producer = await client.NewProducer(Schema.JSON<Product>())
                .Topic(topicName)
                .CreateAsync();

            var consumer = await client.NewConsumer(Schema.JSON<Product>())
                .Topic(topicName)
                .SubscriptionName(subscriptionName)
                .SubscribeAsync();

            var messageId = await producer.SendAsync(new Product{ Name = "IPhone", Rating = 1.1 });
            Console.WriteLine($"MessageId is: '{messageId}'");

            var message = await consumer.ReceiveAsync();
            Console.WriteLine($"Received: {message.GetValue()}");

            await consumer.AcknowledgeAsync(message.MessageId);
        }
    }
}

File: examples/FsharpExamples/Schema.fs
module Schema

// Learn more about F# at http://fsharp.org
open System
open System.ComponentModel.DataAnnotations
open Pulsar.Client.Api

open Pulsar.Client.Common

[<CLIMutable>]
type Product = {
    [<Required>]
    Name: string
    Rating: float
}

let runSchema () =

    let serviceUrl = "pulsar://my-pulsar-cluster:30002"
    let subscriptionName = "my-subscription"
    let topicName = sprintf "my-topic-%i" DateTime.Now.Ticks;

    task {
        
        let! client =
            PulsarClientBuilder()
                .ServiceUrl(serviceUrl)
                .BuildAsync()

        let! producer =
            client.NewProducer(Schema.JSON<Product>())
                .Topic(topicName)
                .CreateAsync()

        let! consumer =
            client.NewConsumer(Schema.JSON<Product>())
                .Topic(topicName)
                .SubscriptionName(subscriptionName)
                .SubscriptionType(SubscriptionType.Exclusive)
                .SubscribeAsync()
        
        let! messageId = producer.SendAsync({ Name = "IPhone"; Rating = 1.1 })
        printfn "MessageId is: '%A'" messageId

        let! message = consumer.ReceiveAsync()
        printfn "Received: %A" (message.GetValue())

        do! consumer.AcknowledgeAsync(message.MessageId)
    }
